### PR TITLE
Raven db 4097 first class patch operations

### PR DESCRIPTION
**Pull request by aviv86**

### CODE REVIEW

File: src/Raven.Client/Documents/Session/DocumentSession.Patch.cs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+            var arg = ((MethodCallExpression)arrayAdder.Body).Arguments[0];
+            LinqPathProvider.GetValueFromExpressionWithoutConversion(arg, out val);
+
+            _documentStore.Operations.Send(new PatchOperation(key, null, new PatchRequest
+            {
+                Script = script,
+                Values = { { "val", val } }
+            }));
+
+            _patchInfo = new { DocId = key, Script = $"\"{script}\"", Path = path, Val = val };
```
Comment (ayende) [inline src/Raven.Client/Documents/Session/DocumentSession.Patch.cs:102]: What if I have multiple calls to Patch? I want all of them.

**Comment by arekpalinski** [comment]
Do we really need to have entire code of `Lambda2JS` lib being a part of RavenDB solution. That's inconsistent with other libs we have: Lucene, Jint are forked by ravendb user (https://github.com/ravendb/) and we keep our code customizations (.net core support) there in separate branches (e.g. https://github.com/ravendb/lucenenet/tree/3.0.3-dnx). RavenDB sln just references DLLs.

**Comment by ayende** [comment]
Yes, we are going to send a PR and then have our own fork

**Comment by arekpalinski** [comment]
OK

File: src/Raven.Client/Documents/Session/DocumentSession.Patch.cs (hunk 1 above)
Comment (ppekrol) [reply src/Raven.Client/Documents/Session/DocumentSession.Patch.cs:102]: Maybe use session.Advanced.Defer ?

File: src/Raven.Client/Documents/Session/DocumentSession.Patch.cs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+            var jsMethodNameAtt = ((MethodCallExpression)arrayAdder.Body).Method.GetCustomAttributes(typeof(JavascriptMethodNameAttribute));
+            var name = ((JavascriptMethodNameAttribute)jsMethodNameAtt.ElementAt(0)).Name;
+
+            var script = name == "concat" ? $"this.{pathScript} = this.{pathScript}{adderScript}" : $"this.{pathScript}{adderScript}";
+
+            object val;
+            var arg = ((MethodCallExpression)arrayAdder.Body).Arguments[0];
+            LinqPathProvider.GetValueFromExpressionWithoutConversion(arg, out val);
+
+            _documentStore.Operations.Send(new PatchOperation(key, null, new PatchRequest
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/DocumentSession.Patch.cs:96]: This should be a part of session not an immediate operation I think?

Comment (ayende) [reply src/Raven.Client/Documents/Session/DocumentSession.Patch.cs:96]: Yes, it should be send as the same tx

**Pull request merged**